aptos-move/framework/supra-framework/sources/automation_registry.move

```diff
@@ -0,0 +1,356 @@
+            &automation_registry.registry_fee_address_signer_cap
+        );
+
+        supra_account::transfer(&resource_signer, to, amount);
+
+        event::emit(FeeWithdrawn { to, amount });
+    }
+
+    /// Update Automation gas limit
+    public entry fun update_automation_gas_limit(
+        supra_framework: &signer,
+        automation_gas_limit: u64
+    ) acquires AutomationRegistry {
+        system_addresses::assert_supra_framework(supra_framework);
+
+        let automation_registry = borrow_global_mut<AutomationRegistry>(@supra_framework);
+        automation_registry.automation_gas_limit = automation_gas_limit;
+
+        event::emit(UpdateAutomationGasLimit { automation_gas_limit });
+    }
+
+    /// Update duration upper limit
+    public entry fun update_duration_upper_limit(
+        supra_framework: &signer,
+        duration_upper_limit: u64
+    ) acquires AutomationRegistry {
+        system_addresses::assert_supra_framework(supra_framework);
+
+        let automation_registry = borrow_global_mut<AutomationRegistry>(@supra_framework);
+        automation_registry.duration_upper_limit = duration_upper_limit;
+
+        event::emit(UpdateDurationUpperLimit { duration_upper_limit });
+    }
+
+    /// Deducts the automation fee from the user's account based on the selected expiry time.
+    fun charge_automation_fee_from_user(owner: &signer, fee: u64) {
+        // todo : dynamic price calculation is pending
+        let registry_fee_address = get_registry_fee_address();
+        supra_account::transfer(owner, registry_fee_address, fee);
+    }
+
+    /// Get last epoch time in second
+    fun get_last_epoch_time_second(): u64 {
+        let last_epoch_time_ms = reconfiguration::last_reconfiguration_time() / MILLISECOND_CONVERSION_FACTOR;
+        last_epoch_time_ms / MILLISECOND_CONVERSION_FACTOR
+    }
+
+    /// Registers a new automation task entry.
+    public entry fun register(
+        owner: &signer,
+        payload_tx: vector<u8>,
+        expiry_time: u64,
+        max_gas_amount: u64,
+        gas_price_cap: u64
+    ) acquires AutomationRegistry {
+        let registry_data = borrow_global_mut<AutomationRegistry>(@supra_framework);
+
+        // todo : well formedness check of payload_tx
+
+        let current_time = timestamp::now_seconds();
+        assert!(expiry_time > current_time, EINVALID_EXPIRY_TIME);
+        assert!((expiry_time - current_time) < registry_data.duration_upper_limit, EEXPIRY_TIME_UPPER);
+
+        let epoch_interval = block::get_epoch_interval_secs();
+        let last_epoch_time = get_last_epoch_time_second();
+        assert!(expiry_time > (last_epoch_time + epoch_interval), EEXPIRY_BEFORE_NEXT_EPOCH);
+
+        registry_data.gas_committed_for_next_epoch = registry_data.gas_committed_for_next_epoch + max_gas_amount;
+        assert!(registry_data.gas_committed_for_next_epoch < registry_data.automation_gas_limit, EGAS_AMOUNT_UPPER);
```

Comment:
I think we should provide means to query `gas_committed_for_next_epoch` in order for user to avoid task registration and being charged for the failed transaction.

Reply:
Okay, sure